### PR TITLE
convert SA controller to shared informers

### PR DESCRIPTION
**Pull request by deads2k**
convert the SA controller to shared informer + workqueue.

I think one of @derekwaynecarr @ncdc or @liggitt

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35230)

### CODE REVIEW

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 1]
```diff
@@ -531,10 +531,11 @@ func StartControllers(s *options.CMServer, kubeconfig *restclient.Config, rootCl
 		}
 	}
 
-	serviceaccountcontroller.NewServiceAccountsController(
+	go serviceaccountcontroller.NewServiceAccountsController(
+		sharedInformers.ServiceAccounts(), sharedInformers.Namespaces(),
 		client("service-account-controller"),
 		serviceaccountcontroller.DefaultServiceAccountsControllerOptions(),
-	).Run()
+	).Run(5, stop)
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/controllermanager.go]: ConcurrentSASyncs

File: pkg/controller/serviceaccount/serviceaccounts_controller.go [hunk 2]
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
...
+	}()
 
-// getServiceAccount returns the ServiceAccount with the given name for the given namespace
-func (e *ServiceAccountsController) getServiceAccount(name, namespace string) (*api.ServiceAccount, error) {
-	key := &api.ServiceAccount{ObjectMeta: api.ObjectMeta{Namespace: namespace}}
-	accounts, err := e.serviceAccounts.Index("namespace", key)
+	ns, err := c.nsLister.Get(key)
 	if err != nil {
-		return nil, err
+		return err
```
Comment (liggitt) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: do we need to tolerate NotFound here? should we add nsSynced to the WaitForCacheSync on startup?

File: pkg/controller/serviceaccount/serviceaccounts_controller.go [hunk 3]
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
...
 
-// getNamespace returns the Namespace with the given name
-func (e *ServiceAccountsController) getNamespace(name string) (*api.Namespace, error) {
-	key := &api.Namespace{ObjectMeta: api.ObjectMeta{Name: name}}
-	namespaces, err := e.namespaces.Index("name", key)
-	if err != nil {
-		return nil, err
-	}
+		if _, err := c.client.Core().ServiceAccounts(ns.Name).Create(&sa); err != nil && !apierrs.IsAlreadyExists(err) {
+			return err
```
Comment (liggitt) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: short-circuits the whole namespace on one failure?

File: pkg/controller/serviceaccount/serviceaccounts_controller.go [hunk 4]
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
...
+		}
 
-	if len(namespaces) == 0 {
-		return nil, nil
 	}
-	if len(namespaces) == 1 {
-		return namespaces[0].(*api.Namespace), nil
+	if err != nil {
```
Comment (liggitt) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: what case is this handling?

File: pkg/controller/serviceaccount/serviceaccounts_controller.go [hunk 5]
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
...
+		case apierrs.IsNotFound(err):
+		case err != nil:
+			return err
 		}
-	}
-	return nil, nil
-}
+		// this is only safe because we never read it and we always write it
+		// TODO eliminate this once the fake client can handle creation without NS
+		sa.Namespace = ns.Name
```
Comment (liggitt) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: yeah, scary looking

File: pkg/controller/serviceaccount/serviceaccounts_controller.go (hunk 3 above)
Comment (deads2k) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: > short-circuits the whole namespace on one failure?

And requeues for retry.  I don't feel strongly though.

File: pkg/controller/serviceaccount/serviceaccounts_controller.go (hunk 2 above)
Comment (deads2k) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: > do we need to tolerate NotFound here?

tolerate no. Special handling to skip. sure.



> should we add nsSynced to the WaitForCacheSync on startup?

I don't see a reason to.  It's what we're making primary choices based on.

File: pkg/controller/serviceaccount/serviceaccounts_controller.go (hunk 4 above)
Comment (deads2k) [inline pkg/controller/serviceaccount/serviceaccounts_controller.go]: > what case is this handling?

just cruft.  Will remove.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 1 above)
Comment (deads2k) [inline cmd/kube-controller-manager/app/controllermanager.go]: > ConcurrentSASyncs

How about I just set it to one until it gets fixed?

**Comment by deads2k** [comment]
Addressed comments.  Left one thread like it was before just to avoid having to do some insane plumbing that we don't really care about for this controller.

**Comment by liggitt** [comment]
LGTM

**Pull request merged**